handbook/python-dataplugins/python-dataplugins.md

```diff
@@ -0,0 +1,310 @@
+            'type': And(str, lambda s: s in (
+               'DataTypeChnFloat32',
+               'DataTypeChnFloat64',
+               'DataTypeChnString',
+               'DataTypeChnDate',
+               'DataTypeChnUInt8',
+               'DataTypeChnInt16',
+               'DataTypeChnInt32',
+               'DataTypeChnInt64'))
+            }]
+      }]}, ignore_extra_keys=True)
+```
+
+All additional "extra keys" will show up as custom properties in DIAdem, Labview,
+or SystemLink DataFinder.
+
+</details>
+
+See full example: [csv-read-with-direct-loading](https://github.com/ni/systemlink-operations-handbook/tree/master/examples/python-dataplugin-examples/csv-read-with-direct-loading)
+
+## Callback Loading
+
+When handling a big data set, you might not want to load all data at the same time.
+To ensure the DataPlugin only returns values that applications request:
+
+   1. Assign an empty array to the channel values property:
+
+      ```python
+      ...
+      'groups': [{
+         ...
+         'channels': [{
+            ...
+            'values': [],
+         }]
+      }]
+      ```
+
+   1. Outsource the functionality to load the bulk data values in a separate function.
+   The function has the following definition:
+
+      ```python
+      def read_channel_values(self, grp_index, chn_index, numberToSkip, numberToTake):
+         """
+            Returns a value array of the correct data type specified in the TDM dictionary
+         """
+      ```
+
+<!-- markdownlint-disable -->
+<details>
+<summary>Example Code</summary>
+<!-- markdownlint-enable -->
+
+```python
+def read_channel_values(self, grp_index, chn_index, numberToSkip, numberToTake):
+   dataType = self.tdm_tree["groups"][grp_index]["channels"][chn_index]["type"]
+   values = []
+   for row in self.data:
+      value = row[self.channelNames[chn_index]]
+      values.append(value)
+   return values[numberToSkip:numberToTake+numberToSkip]
+```
+
+</details>
+
+The client applications call this function to retrieve channel values (or
+a subset of values). You need to implement the function to return the
+values for a given group and channel index. You also need to ensure only the
+correct subset of values is returned for a given `numberToSkip` and `numberToTake`.
+
+In addition, you need to implement a callback function to retrieve the channel
+length. If all channels have the same length, it simply returns a constant.
+
+```python
+def read_channel_length(self, grp_index, chn_index):
+   """
+      Returns the channel length as an Integer for a given group and channel index
+   """
+   return 10
+```
+
+See full example: [csv-read-with-callback-loading](https://github.com/ni/systemlink-operations-handbook/tree/master/examples/python-dataplugin-examples/csv-read-with-callback-loading)
+
+## Error Handling
+
+Python DataPlugins can raise errors in all callback functions. The raised error
+is transported and displayed in DIAdem, LabVIEW, or SystemLink DataFinder.

```

Comment:
Actually, I think it may be sufficient to say that the error is displayed in DIAdem, ... or is transported necessary? Then use "transported to" - my bad !